src/Client/HedgedConnections.cpp

```diff
@@ -0,0 +1,416 @@
+    switch (packet.type)
+    {
+        case Protocol::Server::Data:
+            removeTimeoutsFromReplica(replica, epoll, timeout_fd_to_replica);
+            processReceiveData(replica);
+            addTimeoutToReplica(TimerTypes::RECEIVE_TIMEOUT, replica, epoll, timeout_fd_to_replica, get_hedged_connections.getConnectionTimeouts());
+            break;
+        case Protocol::Server::Progress:
+        case Protocol::Server::ProfileInfo:
+        case Protocol::Server::Totals:
+        case Protocol::Server::Extremes:
+        case Protocol::Server::Log:
+            removeTimeoutFromReplica(TimerTypes::RECEIVE_TIMEOUT, replica, epoll, timeout_fd_to_replica);
+            addTimeoutToReplica(TimerTypes::RECEIVE_TIMEOUT, replica, epoll, timeout_fd_to_replica, get_hedged_connections.getConnectionTimeouts());
+            break;
+
+        case Protocol::Server::EndOfStream:
+            finishProcessReplica(replica, false);
+            break;
+
+        case Protocol::Server::Exception:
+        default:
+            finishProcessReplica(replica, true);
+            break;
+    }
+
+    return packet;
+}
+
+void HedgedConnections::processReceiveData(ReplicaStatePtr & replica)
+{
+    /// When we receive first packet of data from replica, we stop working with replicas, that are
+    /// responsible for the same offset.
+    offsets_with_received_data.insert(replica->parallel_replica_offset);
+
+    for (auto & other_replica : replicas[replica->parallel_replica_offset])
+    {
+        if (other_replica->isReady() && other_replica != replica)
+        {
+            other_replica->connection->sendCancel();
+            finishProcessReplica(other_replica, true);
+        }
+    }
+
+    /// If we received data from replicas with all offsets, we need to stop choosing new replicas.
+    if (get_hedged_connections.hasEventsInProcess() && offsets_with_received_data.size() == replicas.size())
+    {
+        get_hedged_connections.stopChoosingReplicas();
+        if (next_replica_in_process)
+            epoll.remove(get_hedged_connections.getFileDescriptor());
+    }
+}
+
+void HedgedConnections::processTimeoutEvent(ReplicaStatePtr & replica, TimerDescriptorPtr timeout_descriptor)
+{
+    epoll.remove(timeout_descriptor->getDescriptor());
+    replica->active_timeouts.erase(timeout_descriptor->getDescriptor());
+    timeout_fd_to_replica.erase(timeout_descriptor->getDescriptor());
+
+    if (timeout_descriptor->getType() == TimerTypes::RECEIVE_TIMEOUT)
+    {
+        size_t offset = replica->parallel_replica_offset;
+        finishProcessReplica(replica, true);
+
+        /// Check if there is no active connections with the same offset.
+        if (active_connections_count_by_offset[offset] == 0)
+            throw NetException("Receive timeout expired", ErrorCodes::SOCKET_TIMEOUT);
+    }
+    else if (timeout_descriptor->getType() == TimerTypes::RECEIVE_DATA_TIMEOUT)
+    {
+        offsets_queue.push(replica->parallel_replica_offset);
```

Comment:
Can we get this event, e.g., for the second data packet (we cannot start query for other replica in this case)?
For now, it is not clear for me why not.

Reply:
ok, we remove timeout after get first data packet indeed.